File: SpearoGo/Services/MarineService.swift
import Foundation
import CoreLocation

struct MarineService {
    // Open-Meteo marine API — free, no key, global coverage
    // Docs: https://marine-api.open-meteo.com/v1/marine

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.timeoutIntervalForResource = 30
        return URLSession(configuration: config)
    }()

    func fetch(coordinate: CLLocationCoordinate2D) async throws -> MarineData {
        let url = try buildURL(coordinate: coordinate)
        let (data, response) = try await session.data(from: url)

        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw ServiceError.badResponse
        }

        let decoded = try JSONDecoder().decode(OpenMeteoMarineResponse.self, from: data)
        guard let current = decoded.current else { throw ServiceError.missingData }

        return MarineData(
            waveHeight:    current.wave_height ?? 0,
            wavePeriod:    current.wave_period ?? 0,
            waveDirection: current.wave_direction ?? 0,
            seaSurfaceTemp: current.sea_surface_temperature ?? 20,
            fetchedAt:     Date()
        )
    }

    private func buildURL(coordinate: CLLocationCoordinate2D) throws -> URL {
        guard var components = URLComponents(string: Constants.API.marineBase) else {
            throw ServiceError.invalidURL
        }
        components.queryItems = [
            .init(name: "latitude",  value: String(coordinate.latitude)),
            .init(name: "longitude", value: String(coordinate.longitude)),
            .init(name: "current",   value: "wave_height,wave_period,wave_direction,sea_surface_temperature"),
            .init(name: "timezone",  value: "auto")
        ]
        guard let url = components.url else {
            throw ServiceError.invalidURL
        }
        return url
    }
}

// MARK: - Response shapes

private struct OpenMeteoMarineResponse: Decodable {
    let current: CurrentMarine?
}

private struct CurrentMarine: Decodable {
    let wave_height:             Double?
    let wave_period:             Double?
    let wave_direction:          Double?
    let sea_surface_temperature: Double?
}

File: wear/app/src/main/java/com/spearotracker/spearogo/services/MarineService.kt
package com.spearotracker.spearogo.services

import com.spearotracker.spearogo.models.MarineData
import com.spearotracker.spearogo.utils.Constants
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

private interface OpenMeteoMarineApi {
    @GET("v1/marine")
    suspend fun marine(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("current") current: String = "wave_height,wave_period,wave_direction,sea_surface_temperature",
        @Query("timezone") timezone: String = "auto"
    ): OpenMeteoMarineResponse
}

private data class OpenMeteoMarineResponse(
    val current: CurrentMarine?
)

private data class CurrentMarine(
    val wave_height: Double?,
    val wave_period: Double?,
    val wave_direction: Double?,
    val sea_surface_temperature: Double?
)

@Singleton
class MarineService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val api: OpenMeteoMarineApi = Retrofit.Builder()
        .baseUrl(Constants.Api.MARINE_BASE)
        .client(client)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(OpenMeteoMarineApi::class.java)

    suspend fun fetch(latitude: Double, longitude: Double): MarineData {
        val response = api.marine(latitude, longitude)
        val current = response.current ?: throw ServiceException("Missing marine data")

        // Open-Meteo Marine answers HTTP 200 for a landlocked coordinate and
        // returns nulls, so a non-error response is NOT evidence of a sea.
        // Coalescing those nulls to 0.0 and 20.0 is what told a diver in
        // Rheinland-Pfalz the water was a flat calm at 20C, and scored it 9/10.
        val height = current.wave_height
        val temp = current.sea_surface_temperature
        if (height == null || temp == null) throw NoMarineCoverageException()

        return MarineData(
            waveHeight = height,
            wavePeriod = current.wave_period,
            waveDirection = current.wave_direction,
            seaSurfaceTemp = temp
        )
    }
}

/** The coordinate has no sea: the marine API answered, with nothing in it. */
class NoMarineCoverageException : Exception("No marine coverage for this location")
